scipy/stats/stats.py

```diff
@@ -4361,13 +4358,21 @@ def ks_2samp(data1, data2):
     """
     data1 = np.sort(data1)
     data2 = np.sort(data2)
-    n1 = data1.shape[0]
-    n2 = data2.shape[0]
-    data_all = np.concatenate([data1, data2])
-    cdf1 = np.searchsorted(data1, data_all, side='right') / (1.0*n1)
-    cdf2 = np.searchsorted(data2, data_all, side='right') / (1.0*n2)
-    d = np.max(np.absolute(cdf1 - cdf2))
-    # Note: d absolute not signed distance
+    n1, = data1.shape
+    n2, = data2.shape
+    common_type = np.find_common_type([], [data1.dtype, data2.dtype])
+    if not (np.issubdtype(common_type, np.number) and
+            not np.issubdtype(common_type, np.complexfloating)):
+        raise ValueError('ks_2samp only accepts real inputs')
+    # nans, if any, are at the end after sorting.
+    if np.isnan(data1[-1]) or np.isnan(data2[-1]):
+        raise ValueError('ks_2samp only accepts non-nan inputs')
```

Comment:
This is, strictly speaking, a back-compat break, so it needs a mention in the release notes.